src/zkevm_specs/evm/execution/end_block.py

```diff
@@ -53,18 +54,29 @@ def end_block(instruction: Instruction):
             ]
         )
     )
+    # total_valid_txs = total_txs - invalid_txs
+    total_valid_txs = FQ(
+        total_txs
+        - len(
+            [
+                tx_row
+                for tx_row in instruction.tables.tx_table
+                if tx_row.field_tag == TxContextFieldTag.TxInvalid and tx_row.value.expr() == FQ(1)
+            ]
+        )
+    )
     # Note that rw_counter starts at 1
     is_empty_block = instruction.is_zero(instruction.curr.rw_counter - 1)
     # If the block is not empty, we will do 1 call_context lookup
     total_rws = (1 - is_empty_block) * (instruction.curr.rw_counter - 1 + 1)
 
     if instruction.is_last_step:
-        # 1. Constraint total_txs witness values depending on the empty block case.
+        # 1. Constraint total_valid_txs witness values depending on the empty block case.
         if is_empty_block == FQ(1):
-            # 1a. total_txs is 0 in empty block
-            instruction.constrain_equal(total_txs, FQ(0))
+            # 1a. total_valid_txs is 0 in empty block
+            instruction.constrain_equal(total_valid_txs, FQ(0))
         else:
-            # 1b. total_txs matches the tx_id that corresponds to the final step.
+            # 1b. total_valid_txs matches the tx_id that corresponds to the final step.
             instruction.constrain_equal(
                 instruction.call_context_lookup(CallContextFieldTag.TxId), total_txs
```

Comment:
```suggestion
                instruction.call_context_lookup(CallContextFieldTag.TxId), total_valid_txs
```
As the comment above says?

Reply:
no, the comments is replaced by mistake, here still using total_txs.
That is part of the reason why I wanted to use another new table in front of the existed tx table (if possible/necessary), as we need to carefully deal with existed tx table log now, and later circuit writers need to keep the invalid skip logic in mind once this flag is added. Not sure if it add extra burden to others, however, so far these changes seems workable and acceptable I guess?? What's your opinion?

Reply:
Oh okay. Is the only difficulty recovering the "real" tx id (counting up using only the valid transactions)? If so seems easy enough to keep track of that (and perhaps even put it in the tx table).

Reply:
Yes, I now think adding more flags (auxiliary table columns) to tx table if necessary is a better approach, although those who query tx table need to keep them in mind and adjust the way how to use the values accordingly.